### PR TITLE
[WIP] add more rrules for imul!

### LINKED ISSUE #22: add rrule in complex cases
All the rrules in real cases have been added and now I want to think about the complex and adjoint cases.
However, It seems `FiniteDifferences.jl` doesn't support the jv'p in complex cases. When I just initialize the nzval as random complex values, it would went wrong as follows,
``` julia
Got exception outside of a @test
  MethodError: no method matching _j′vp(::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, ::ComposedFunction{ComposedFunction{ComposedFunction{typeof(first), typeof(FiniteDifferences.to_vec)}, FiniteDifferences.var"#79#80"{ChainRulesTestUtils.var"#fnew#42"{ChainRulesTestUtils.var"#call#50"{NamedTuple{(), Tuple{}}}, Tuple{typeof(*), SparseMatrixCSC{ComplexF64, Int64}, Matrix{ComplexF64}}, Tuple{Bool, Bool, Bool}}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{var"#SparseMatrixCSC_from_vec#1"{SparseMatrixCSC{ComplexF64, Int64}, SparseMatrixCSC{ComplexF64, Int64}}, FiniteDifferences.var"#Array_from_vec#34"{Matrix{ComplexF64}, FiniteDifferences.var"#Vector_from_vec#32"{Vector{ComplexF64}, Vector{FiniteDifferences.var"#Complex_from_vec#21"}, Vector{Vector{Float64}}}}}}}, ::Vector{Float64}, ::Vector{ComplexF64})
  Closest candidates are:
    _j′vp(::Any, ::Any, ::Vector{var"#s16"} where var"#s16"<:Real, ::Vector{var"#s15"} where var"#s15"<:Real) at /Users/lijie/.julia/packages/FiniteDifferences/W3rQO/src/grad.jl:78
  Stacktrace:
```

So I just try another way: concat the real part and imag part as nzval, related codes are as follows:

``` julia
# ChainRulesTestUtils requires some method override to support sparse array
# BEGIN ChainRulesTestUtils

# we use this to generate Tangent and pass to test_rrule
# for example: A ⊢ sprand_tangent(A)
function sprand_tangent(A::SparseMatrixCSC{Tv}) where Tv<:Real
    Ā = copy(A)
    Ā.nzval .= rand(eltype(A), length(A.nzval))
    return Tangent{SparseMatrixCSC}(Ā)
end

function sprand_tangent(A::SparseMatrixCSC{Tv}) where Tv<:Complex
    Ā = SparseMatrixCSC(
        A.m,
        A.n,
        A.colptr,
        A.rowval,
        rand(real(Tv), 2*length(A.nzval))
    )
    return Tangent{SparseMatrixCSC}(Ā)
end

function Base.:(+)(a::P, b::Tangent{P}) where P<:SparseMatrixCSC
    b[1] .+ a
    return b
end

function FiniteDifferences.to_vec(A::SparseMatrixCSC{Tv}) where Tv<:Real
    v = A.nzval
    function SparseMatrixCSC_from_vec(v)
        m = A.m
        n = A.n
        colptr = A.colptr
        rowval = A.rowval
        SparseMatrixCSC(m, n, colptr, rowval, v)
    end
    return v, SparseMatrixCSC_from_vec
end

function FiniteDifferences.to_vec(A::SparseMatrixCSC{Tv}) where Tv<:Complex
    v = collect(reinterpret(real(Tv), A.nzval))
    function SparseMatrixCSC_from_vec(v)
        m = A.m
        n = A.n
        colptr = A.colptr
        rowval = A.rowval
        SparseMatrixCSC(m, n, colptr, rowval, v)
    end
    return v, SparseMatrixCSC_from_vec
end
# END ChainRulesTestUtils
```

and then test it 

``` julia
@testset "test rrule" begin
    @testset "matrix multiplication" begin
        # sparse * vector
        for T in (Float64, ComplexF64)
            A = sprand(T, 5, 3, 0.5)
            B = rand(T, 3)
            test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
            # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.

            # sparse * dense matrix
            A = sprand(T, 5, 3, 0.5)
            B = rand(T, 3, 2)
            test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
            # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.
        end 
    end
end
```
Complex cases went wrong with error messages as follows:

``` julia
test_rrule: * on SparseMatrixCSC{ComplexF64, Int64},Vector{ComplexF64}: Test Failed at /Users/lijie/.julia/packages/ChainRulesTestUtils/ugS4W/src/check_result.jl:24
  Expression: isapprox(actual, expected; kwargs...)
   Evaluated: isapprox(
           ⋅                     ⋅                   ⋅    
 0.0389629-0.364324im            ⋅           1.36762-2.9693im
 -0.696108-0.37008im             ⋅                   ⋅    
  0.370505-0.627416im  -0.385663-0.815565im  4.92051-4.24894im
  0.542053-0.102961im   0.271014-0.627069im  4.87641+0.674131im, 
   ⋅           ⋅         ⋅ 
  0.0389629    ⋅        1.36762
 -0.696108     ⋅         ⋅ 
  0.370505   -0.385663  4.92051
  0.542053    0.271014  4.87641; rtol = 1.0e-9, atol = 1.0e-9)
Stacktrace:
 [1] test_approx(actual::SparseMatrixCSC{ComplexF64, Int64}, expected::SparseMatrixCSC{Float64, Int64}, msg::String; kwargs::Base.Iterators.Pairs{Symbol, Float64, Tuple{Symbol, Symbol}, NamedTuple{(:rtol, :atol), Tuple{Float64, Float64}}})
   @ ChainRulesTestUtils ~/.julia/packages/ChainRulesTestUtils/ugS4W/src/check_result.jl:24
 [2] macro expansion
   @ ~/.julia/packages/ChainRulesTestUtils/ugS4W/src/testers.jl:238 [inlined]
 [3] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [4] test_rrule(::ChainRulesTestUtils.ADviaRuleConfig, ::typeof(*), ::ChainRulesTestUtils.PrimalAndTangent{SparseMatrixCSC{ComplexF64, Int64}, Tangent{SparseMatrixCSC, Tuple{SparseMatrixCSC{Float64, Int64}}}}, ::Vararg{Any, N} where N; output_tangent::ChainRulesTestUtils.Auto, check_thunked_output_tangent::Bool, fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, rrule_f::Function, check_inferred::Bool, fkwargs::NamedTuple{(), Tuple{}}, rtol::Float64, atol::Float64, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ChainRulesTestUtils ~/.julia/packages/ChainRulesTestUtils/ugS4W/src/testers.jl:194
```

The real part is totally right. Here I want to metion that the initilization may be wrong:
``` julia
function sprand_tangent(A::SparseMatrixCSC{Tv}) where Tv<:Complex
    Ā = SparseMatrixCSC(
        A.m,
        A.n,
        A.colptr,
        A.rowval,
        rand(real(Tv), 2*length(A.nzval))
    )
    return Tangent{SparseMatrixCSC}(Ā)
end
```
it just return `Ā.nzval `  same with the `real(Ā.nzval)`. That is the reason why the real part is totally right. We should find a way to insert imag part into gradient and convert the expected type into complex SparseCSC (must be same with the primal type).

cc: @GiggleLiu @johnnychen94

### PR DESCRIPTION
**Pull request by jieli-matrix**
In this PR, I want to complete these work:
 - [x] add all rrules for imul!
 - [ ] test rrules using test_rrule (complete partially)

### CODE REVIEW

File: test/chainrules.jl [hunk 1]
```diff
@@ -1,15 +1,35 @@
...
+        #end 
+    end
+
+    @testset "adjoint dense matrix-sparse matrix multiplication" begin
+    #     # adjoint dense matrix * sparse matrix
+    #     for T in (Float64)
+            B = rand(10, 5)
+            A = sprand(10, 5, 0.5)
+            test_rrule(*, B', A ⊢ sprand_tangent(A); check_thunked_output_tangent=false)
+    #     end
```
Comment (jieli-matrix) [inline test/chainrules.jl:33]: It seems nothing errored when I do a simple test on my local machine:
``` julia
B = rand(10, 5)
A = sprand(10, 5, 0.5)
C, C_pullback = ChainRulesCore.rrule(*, B', A)
C̄ = rand(5,5)
_,  B̄, Ā = C_pullback(C̄)
```
But the test fails
``` julia
return type Nothing does not match inferred return type Union{}
```
But it truly return in my simple test case, maybe it fails in a specific example?

File: test/chainrules.jl [hunk 2]
```diff
@@ -1,15 +1,35 @@
...
+    end
+
+    @testset "dense matrix-sparse matrix multiplication" begin
+        # dense matrix * sparse matrix
+        #for T in (Float64)
+            B = rand(10, 10)
+            A = sprand(10, 5, 0.5)
+            test_rrule(*, B, A ⊢ sprand_tangent(A); check_thunked_output_tangent=false)
+        #end 
+    end
```
Comment (jieli-matrix) [inline test/chainrules.jl:25]: I think maybe complex case should be added into rrule. But it really confused me, when I tried to write as follows:
``` julia
for T in (Float64)
            B = rand(T, 10, 10)
            A = sprand(T, 10, 5, 0.5)
            test_rrule(*, B, A ⊢ sprand_tangent(A); check_thunked_output_tangent=false)
end 
```
It went wrong and told me
``` julia
Got exception outside of a @test
  MethodError: no method matching iterate(::Type{Float64})
  Closest candidates are:
    iterate(::Union{LinRange, StepRangeLen}) at range.jl:664
    iterate(::Union{LinRange, StepRangeLen}, ::Int64) at range.jl:664
    iterate(::T) where T<:Union{Base.KeySet{var"#s79", var"#s78"} where {var"#s79", var"#s78"<:Dict}, Base.ValueIterator{var"#s77"} where var"#s77"<:Dict} at dict.jl:693
    ...
```
Why? Is there any differences?

File: src/chainrules.jl [hunk 3]
```diff
@@ -1,8 +1,43 @@
...
+function ChainRulesCore.rrule(
+    ::typeof(*), xA::Adjoint{T, <:AbstractSparseMatrix}, B::DenseInputVecOrMat
+) where T
+    C = xA * B
+    function mul_pullback(C̄)
+        _, i_xA, i_B, _, _ = (~imul!)(AD.GVar(C, C̄), AD.GVar(xA), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
+        return ChainRulesCore.NoTangent(), AD.grad(i_xA), AD.grad(i_B)
+    end
+    return C, mul_pullback
+end
```
Comment (jieli-matrix) [inline src/chainrules.jl:21]: After we finished the complex case, we need to think about adjoint of sparsecsc. Maybe a new sprand_tangent method is needed?

**Comment by jieli-matrix** [comment]
As soon as we agree on complex and adjoint cases are necessary in rrule, I would work on tasks as follows:
- [ ] `function sprand_tangent(A::AT) where AT<:SparseMatrixCSC{:<Complex}`
- [ ] `function FiniteDifferences.to_vec(A::AT) where AT<:SparseMatrixCSC` (in complex cases, only convert v is enough)
- [ ] `function sprand_tangent(A::AT) where AT<:Adjoint{SparseMatrixCSC}`
- [ ]  `function FiniteDifferences.to_vec(A::AT) where AT<:Adjoint{SparseMatrixCSC}`
Besides, all the related tests would be added.

File: test/chainrules.jl (hunk 1 above)
Comment (jieli-matrix) [reply test/chainrules.jl:33]: This problem has already been solved, ignore it.

File: src/chainrules.jl (hunk 3 above)
Comment (GiggleLiu) [reply src/chainrules.jl:21]: I am not sure about this. I think the adjoint of the Adjoint type should also be something like Adjoint. However, many people define the adjoint of a dense matrix as a normal matrix.

File: test/chainrules.jl (hunk 2 above)
Comment (GiggleLiu) [reply test/chainrules.jl:25]: In julia 1.7, the error message is more intuitive:
```julia
ERROR: MethodError: no method matching iterate(::Type{Float64})
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at range.jl:737
  iterate(::Union{LinRange, StepRangeLen}, ::Int64) at range.jl:737
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}} at dict.jl:695
  ...
Stacktrace:
 [1] top-level scope
   @ ./REPL[6]:1
```

`(Float64)` This is a type, `(Float64,)` is a tuple.

File: src/chainrules.jl (hunk 3 above)
Comment (jieli-matrix) [reply src/chainrules.jl:21]: May be our implementation should refer to `ProjectTo` in ChainRulesCore. This is [sparsecsc implementation](https://github.com/JuliaDiff/ChainRulesCore.jl/blob/1014715a38a0f5a9c1e84a804fb1b289e7fb760f/src/projection.jl#L440) and [adjoint implementation](https://github.com/JuliaDiff/ChainRulesCore.jl/blob/1014715a38a0f5a9c1e84a804fb1b289e7fb760f/src/projection.jl#L440). I'm not sure if both of the methods were added, would it adapt to `Adjoint{SparseCSC}`? Or we should implement `sprand_tangent` similar to `ProjectTo` on `Adjoint{SparseCSC}` type?

File: test/chainrules.jl (hunk 2 above)
Comment (jieli-matrix) [reply test/chainrules.jl:25]: Thx!!!

Comment (jieli-matrix) [reply test/chainrules.jl:25]: After I corrected the syntax error,  I guess may be I need to add to_vec for complex case when I see the error message:

``` julia
Got exception outside of a @test
  MethodError: no method matching _j′vp(::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, ::ComposedFunction{ComposedFunction{ComposedFunction{typeof(first), typeof(FiniteDifferences.to_vec)}, FiniteDifferences.var"#79#80"{ChainRulesTestUtils.var"#fnew#42"{ChainRulesTestUtils.var"#call#50"{NamedTuple{(), Tuple{}}}, Tuple{typeof(*), SparseMatrixCSC{ComplexF64, Int64}, Matrix{ComplexF64}}, Tuple{Bool, Bool, Bool}}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{var"#SparseMatrixCSC_from_vec#1"{SparseMatrixCSC{ComplexF64, Int64}, SparseMatrixCSC{ComplexF64, Int64}}, FiniteDifferences.var"#Array_from_vec#34"{Matrix{ComplexF64}, FiniteDifferences.var"#Vector_from_vec#32"{Vector{ComplexF64}, Vector{FiniteDifferences.var"#Complex_from_vec#21"}, Vector{Vector{Float64}}}}}}}, ::Vector{Float64}, ::Vector{ComplexF64})
  Closest candidates are:
    _j′vp(::Any, ::Any, ::Vector{var"#s16"} where var"#s16"<:Real, ::Vector{var"#s15"} where var"#s15"<:Real) at /Users/lijie/.julia/packages/FiniteDifferences/W3rQO/src/grad.jl:78
  Stacktrace:
```

**Comment by jieli-matrix** [comment]
Already solved by #24

**Pull request closed without merging**